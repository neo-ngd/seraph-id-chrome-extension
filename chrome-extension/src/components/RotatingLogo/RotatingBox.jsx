// Copyright (c) 2019 Swisscom Blockchain AG
// Licensed under MIT License

import React, { Component } from 'react';
import PropTypes from 'prop-types';

/**
 * <React3DRotationBox />
 * 3D rotation box. It rotates the passed element based on the mouse position.
 */
class React3DRotationBox extends Component {
  constructor(props) {
    super(props);

    this.state = {
      rotateXDeg: 0,
      rotateYDeg: 0,
    };

    this._moveBox = this._moveBox.bind(this);
  }

  _moveBox(e) {
    const rotateForce = this.props.rotateForce;
    const docWidth = this.docWidth;
    const docHeight = this.docHeight;

    // rotate range: -rotateForce ~ rotateForce
    this.setState({
      rotateXDeg: -(((e.pageY / docHeight) * 2 - 1) * rotateForce),
      rotateYDeg: ((e.pageX / docWidth) * 2 - 1) * rotateForce,
    });
  }

  componentDidMount() {
    this.docWidth = document.body.offsetWidth;
    this.docHeight = document.body.offsetHeight;
    document.body.addEventListener('mousemove', this._moveBox);
  }

  componentWillUnmount() {
    document.body.removeEventListener('mousemove', this._moveBox);
  }

  render() {
    const props = this.props;
    return (
      <div
        className={`motion-box ${this.props.className || ''}`}
        style={{
          perspective: props.perspective,
          minWidth: props.minWidth,
          minHeight: props.minHeight,
          textAlign: 'center',
        }}
      >
        <div
          style={{
            transformStyle: 'preserve-3d',
            transform: `rotateX(${this.state.rotateXDeg}deg) rotateY(${this.state.rotateYDeg}deg)`,
          }}
        >
          {this.props.children}
        </div>
      </div>
    );
  }
}

React3DRotationBox.defaultProps = {
  rotateForce: 20,
  perspective: '300px',
  minWidth: 'auto',
  minHeight: '150px',
};

React3DRotationBox.propTypes = {
  rotateForce: PropTypes.number,
  perspective: PropTypes.string,
  minWidth: PropTypes.string,
  minHeight: PropTypes.string,
};

export default React3DRotationBox;
