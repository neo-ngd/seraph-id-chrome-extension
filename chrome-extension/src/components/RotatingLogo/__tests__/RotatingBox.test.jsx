import React from 'react';
import { shallow } from 'enzyme';
import RotatingBox from "../RotatingBox";


describe('<RotatingBox />', () => {
    it('should render component', () => {
        const wrapper = shallow(<RotatingBox />);
        expect(wrapper).toMatchSnapshot()
    });
});
