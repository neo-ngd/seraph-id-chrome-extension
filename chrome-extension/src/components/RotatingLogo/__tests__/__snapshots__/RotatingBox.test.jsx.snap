// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RotatingBox /> should render component 1`] = `
<div
  className="motion-box "
  style={
    Object {
      "minHeight": "150px",
      "minWidth": "auto",
      "perspective": "200px",
      "textAlign": "center",
    }
  }
>
  <div
    style={
      Object {
        "transform": "rotateX(0deg) rotateY(0deg)",
        "transformStyle": "preserve-3d",
      }
    }
  />
</div>
`;
